nwbinspector/utils.py

```diff
@@ -0,0 +1,26 @@
+"""Authors: Cody Baker and Ben Dichter."""
+import numpy as np
+from collections import defaultdict
+
+global default_checks
+default_checks = {1: defaultdict(list), 2: defaultdict(list), 3: defaultdict(list)}
```

Comment:
Hardcoding the ``severity`` level to ``[1,2,3]`` seems problematic from a maintenance perspective because it makes it hard to add additional levels and it is not explicit what these levels mean.

Reply:
The idea being we would never add additional levels beyond this; the mapping we were thinking (also in terms of nicer output display) is `1=low=green`, `2=moderate=yellow`, `3=severe=red`, and all future checks would be classified by this.

Granted, we could just as well make those three text-based options our actual severity levels so that it's not implied that it can take any positive integer value. The main reason we started that way was to make the `severity_threshold` of the main function easier to evaluate, but we could also just enumerate over the ordered text levels to the same effect.

Reply:
I think this approach to severity may ultimately be a bit too limiting. Let me just throw out a few ideas:

1. Currently it looks that each check just returns a string. This is limiting because it means each check can only have two return values, i.e, 1) pass or 2) fail with a predetermined level of severity that is set in the decoratory. Instead, I think it would be useful to have checks return error/warning classes (e.g., a DimensionError if dimensions of an error are incorrect). The NWBInspector could define its own base class for errors and warnings, which would allow you to then also associate custom properties with different error types (e.g., the color for printing, severity etc.. This would:
    * make it easier in the future extend the approach to define new error types if necessary (rather than being limited to just 3 levels of severity plus a name for the check)
    *  make filtering, sorting, and formatting of error messages easier after you have completed all checks as the properties (e.g., severity, color etc.) would be associated with each error message that was generated, rather than being properties of the check that generated the error
    * allow checks to generate multiple different types of error message, e.g.,the check ``check_dtype`` may find that the dtype is not allowed and, hence, create an error or it may find that a different dtype may be more appropriate and as such only create a warning
    * allow a check to generate multiple errors, instead of being limited to a single message
    * allow  you to distinguish between the importance of performing a check and the criticality of the error it may have generate
2. In addition to (or instead of) ``severity``, I think it would be useful to assign a ``catagory`` label for checks
3. Instead of hard-coding properties of checks (e.g., the severity) in the dict itself, I think it may be simpler to make these attributes of the checks themselves (e.g., by wrapping the check functions in a class). In this way the ``default_checks`` dict could turn into a simple list. In this way, properties are always associated with the objects, rather than being managed outside, so that you can easily filter checks based on their properties, e.g., their severity or category
4. For checks, I think the term ``importance`` may be more appropriate than ``severity``. The term ``severity`` appears to be more appropriate for the actual error a check may have generated rather than the check itself.

Reply:
@oruebel 

Thanks for helping us determine the structure of this tool. The numeric values for `severity` took inspiration from the number codes in the Python `logger` library ([doc](https://docs.python.org/3/howto/logging.html#logging-levels)). I see what you mean that having function return error types offers flexibility of 1) allowing a check to output different types of errors, and 2) allowing us to expand upon the types of errors. As for 1, we haven't seen the need for this yet, and I think in most cases that can easily be refactored into two different checks. As for 2) I actually don't think those different types of errors will be as useful as severity. It will be of no consequence to the user if they got a `DimensionalityError` vs other errors or warnings. What they want to know is whether or not they _need_ to fix it. I really want to avoid over-engineering here, creating categories and labels that are never used. After experience reviewing dozens of different datasets from all different kinds of labs, the most important information we can attached to each message is how important it is. This is a key idea because we may produce dozens of warnings but only one or two are really essential, and need to take priority. I suppose we could try to map the errors with levels or labels to indicate this, but then we'd still get back to the need to implement a mapping to some sort of severity scale in the end anyhow and it's not clear that the error types would necessarily match deterministically to severity. For instance, you could have a severe dimensionality warning, e.g. if TimeSeries.data is transposed, or a non-severe dimensionality warning, e.g. if you have a dataset that is shape [N,1,1] when it could be just [N,]. 

Annotating the checks with severity has the advantage that it allows us to skip checks based on severity, which would be particularly useful if some of the less severe checks take a long time to run.

> make filtering, sorting, and formatting of error messages easier after you have completed all checks as the properties (e.g., severity, color etc.) would be associated with each error message that was generated, rather than being properties of the check that generated the error

I don't see how this would be any easier than what we are currently doing.

> check_dtype may find that the dtype is not allowed

This should not happen, since that would be caught be the validator. I want to make sure that we don't confusing the scopes of these two different tools.

> allow a check to generate multiple errors, instead of being limited to a single message

These functions can output different strings. They would be locked to the same severity though, this is true.

As for the question of decorator vs. class, I've been considering both. In these types of situations, I like to take inspiration from the most analogous established tool in the Python ecosystem. In this case I think we are doing something similar to testing, but Python has pytest, which is function-based, and unittest, which is class-based, so that doesn't help us here. I am leaning towards decorator. Note that you would still be able to attached attributes to functions similarly to how you do with classes. See the code below for setting an attribute of a function.

```python
def nwbinspector_check(severity: int):
    """Wrap a check function to add it to the list of default checks for that severity and neurodata type."""

    def decorator(check_function):
        check_function.severity = severity
        return check_function

    return decorator


@nwbinspector_check(1)
def check_pynwb(nwbfile):
    pass

check_pynwb.severity == 1
```

I don't see any difference in functionality between the two approaches, so the decorator approach is nicer to me simply because it's less code. 

> allow you to distinguish between the importance of performing a check and the criticality of the error it may have generate

I disagree with the idea of distinguishing between the importance of a check and the severity of the error. It complicates things in a bad way that can only lead to trouble. If a check is capable of producing a severe warning, it is an important check. Any other designation would be an opportunity to miss something important. I suppose we could go down the path of having checks output warnings of different severity, but in that case I would want the level of the check to reflect the highest level of warning it is capable of producing.

Reply:
I think it may be worth having a chat next week to go over these items